### PR TITLE
feat: add DefaultFunc for bytes and strings

### LINKED ISSUE #935: proposal: add support for DefaultFunc for string fields
Is it an accepted idea to add new build-in custom id type - cuid?

https://github.com/ericelliott/cuid

### PR DESCRIPTION
**Pull request by nolotz**
This adds default funcs for string and bytes builders.

Fixes #935 also.

### CODE REVIEW

**Comment by a8m** [comment]
Thanks for your contribution @nolotz, it's appreciated!

File: entc/gen/type.go [hunk 1]
```diff
@@ -67,6 +67,8 @@ type (
 		Nillable bool
 		// Default indicates if this field has a default value for creation.
 		Default bool
+		// DefaultFunc indicates if the fields default value is a function and should be executed.
+		DefaultFunc bool
```
Comment (a8m) [inline entc/gen/type.go:71]: Can you please replace it with the following method:

```go
func (f Field) DefaultFunc() bool { return f.def.DefaultKind == reflect.Func }
```

File: entc/gen/type.go [hunk 2]
```diff
@@ -199,6 +201,7 @@ func NewType(c *Config, schema *load.Schema) (*Type, error) {
 			Nillable:      f.Nillable,
 			Optional:      f.Optional,
 			Default:       f.Default,
+			DefaultFunc:   f.DefaultFunc,
```
Comment (a8m) [inline entc/gen/type.go:204]:
```suggestion
			Default:       f.Default,
```

See comment above.

File: entc/load/schema.go [hunk 3]
```diff
@@ -46,6 +46,7 @@ type Field struct {
 	Nillable      bool                    `json:"nillable,omitempty"`
 	Optional      bool                    `json:"optional,omitempty"`
 	Default       bool                    `json:"default,omitempty"`
+	DefaultFunc   bool                    `json:"defaultFunc,omitempty"`
```
Comment (a8m) [inline entc/load/schema.go:49]:
```suggestion
	Default       bool                    `json:"default,omitempty"`
```

See comment above.

File: entc/load/schema.go [hunk 4]
```diff
@@ -116,6 +117,7 @@ func NewField(fd *field.Descriptor) (*Field, error) {
 		Nillable:      fd.Nillable,
 		Optional:      fd.Optional,
 		Default:       fd.Default != nil,
+		DefaultFunc:   fd.DefaultFunc,
```
Comment (a8m) [inline entc/load/schema.go:120]:
```suggestion
```

File: schema/field/field.go [hunk 5]
```diff
@@ -892,6 +908,7 @@ type Descriptor struct {
 	Optional      bool                    // nullable field in database.
 	Immutable     bool                    // create-only field.
 	Default       interface{}             // default value on create.
+	DefaultFunc   bool                    // default value is a function.
```
Comment (a8m) [inline schema/field/field.go:911]:
```suggestion
```

File: schema/field/field.go [hunk 6]
```diff
@@ -205,6 +205,14 @@ func (b *stringBuilder) Validate(fn func(string) error) *stringBuilder {
...
+	b.desc.DefaultFunc = false
+	return b
+}
+
+// DefaultFunc sets the default value generating function of the field.
+func (b *stringBuilder) DefaultFunc(fn func() string) *stringBuilder {
+	b.desc.Default = fn
+	b.desc.DefaultFunc = true
 	return b
 }
```
Comment (a8m) [inline schema/field/field.go:217]:
```suggestion
// DefaultFunc sets the function that is applied to set default value
// of the field on creation. For example:
//
//	field.String("cuid").
//		DefaultFunc(cuid.New)
//
func (b *stringBuilder) DefaultFunc(fn func() string) *stringBuilder {
	b.desc.Default = fn
	return b
}
```

File: schema/field/field.go [hunk 7]
```diff
@@ -481,6 +489,14 @@ type bytesBuilder struct {
...
+	b.desc.DefaultFunc = false
+	return b
+}
+
+// Default sets the default value of the field.
+func (b *bytesBuilder) DefaultFunc(v func() []byte) *bytesBuilder {
+	b.desc.Default = v
+	b.desc.DefaultFunc = true
 	return b
 }
```
Comment (a8m) [inline schema/field/field.go:501]:
```suggestion
// Default sets the default value of the field.
func (b *bytesBuilder) DefaultFunc(v func() []byte) *bytesBuilder {
	b.desc.Default = v
	return b
}
```

Update also the godoc as mentioned in the previous comment.

File: schema/field/field_test.go [hunk 8]
```diff
@@ -183,6 +183,17 @@ func TestBytes(t *testing.T) {
 	assert.True(t, fd.Info.Nillable)
 	assert.True(t, fd.Info.ValueScanner())
 
+	fd = field.Bytes("uuid").GoType(&uuid.UUID{}).DefaultFunc(func() []byte {
+		return []byte("{}")
+	}).Descriptor()
```
Comment (a8m) [inline schema/field/field_test.go:188]:
```suggestion
	fd = field.Bytes("uuid").
		GoType(&uuid.UUID{}).
		DefaultFunc(func() []byte {
			return []byte("{}")
		}).
		Descriptor()
```

File: schema/field/field_test.go [hunk 9]
```diff
@@ -192,9 +203,23 @@ func TestBytes(t *testing.T) {
...
 func TestString(t *testing.T) {
+	fd := field.String("name").Default("Ent").Descriptor()
+	assert.Equal(t, "name", fd.Name)
+	assert.Equal(t, field.TypeString, fd.Info.Type)
+	assert.NotNil(t, fd.Default)
+	assert.Equal(t, "Ent", fd.Default)
+
+	fd = field.String("name").DefaultFunc(func() string {
+		return "Ent"
+	}).Descriptor()
```
Comment (a8m) [inline schema/field/field_test.go:214]:
```suggestion
	fd = field.String("name").
		DefaultFunc(func() string {
			return "Ent"
		}).
		Descriptor()
```

**Comment by nolotz** [comment]
Thanks for the code review @a8m.

Next time please warn me, that you already implemented the reflection part ;)

It would be awesome when you also take a look at #1085.

best regards

**Comment by a8m** [comment]
> Thanks for the code review @a8m.
> 
> Next time please warn me, that you already implemented the reflection part ;)

Hehe, sorry for that. I wanted to unblock you 😄 

> It would be awesome when you also take a look at #1085.

I'll give it a look later today or tomorrow morning. Thanks a lot for the great work!

**Review by a8m — APPROVED** [review]
🥳

**Pull request merged**